File: sigi/lexer.py
"""Sigi lexer - pure symbolic tokenization."""

from dataclasses import dataclass
from typing import List, Optional, Union


@dataclass
class Token:
    kind: str
    value: Optional[Union[float, int, str]] = None
    line: int = 1
    col: int = 1

    def __repr__(self):
        if self.value is not None:
            return f"Token({self.kind!r}, {self.value!r})"
        return f"Token({self.kind!r})"


class LexError(Exception):
    pass


class Lexer:
    """
    Sigi lexer - all syntax is symbolic.

    Symbol reference:
    !N      Push number N (int or float)
    @       DUP - duplicate top
    #       SWAP - swap top two
    $       DROP - discard top
    +       ADD
    -       SUB
    *       MUL
    /       DIV
    %       MOD
    =       EQ - push 1 if equal, else 0
    <       LT - push 1 if a < b
    >       GT - push 1 if a > b
    ~       NOT - push 1 if 0, else 0
    |       Print number with newline
    ^       Print as character
    ?       Read number from input
    :       Store (pop val, pop addr, store to var)
    .       Load (pop addr, push var value)
    <n>     Push variable n directly (0-99)
    [ ]     WHILE loop - pop condition, loop while nonzero
    (N)     Call function N (0-99)
    { }     Block - for function definitions and conditionals
    ;       ELSE separator in blocks
    "text"  String literal
    'x      Character literal (no closing ')
    \\      End of line comment
    S       SIN - pop a, push sin(a)
    C       COS - pop a, push cos(a)
    T       TAN - pop a, push tan(a)
    R       SQRT - pop a, push sqrt(a)
    P       POW - pop exp, pop base, push pow(base, exp)
    F       FLOOR - pop a, push floor(a)
    L       LOG - pop a, push log(a)
    E       EXP - pop a, push exp(a)
    M       ABS - pop a, push fabs(a)
    N       ATAN2 - pop y, pop x, push atan2(y, x)
    W       RAND - push random value 0.0-1.0
    X       EXIT - pop code, exit program
    Z       TIME - push current time in seconds
    &       ALEN - pop arr_id, push array length
    A       ALOAD - pop idx, pop arr_id → push array[idx]
    a       ASTORE - pop val, pop idx, pop arr_id → array[idx] = val
    _       AINIT - pop arr_id, pop size → initialize array
    U       USLEEP - pop microseconds, sleep

    Tokens:
    NUM     - numeric value (from ! prefix)
    VAR     - variable index (digits directly)
    CHAR    - character literal
    STRING  - string literal
    BLOCK   - { start block
    ENDB    - } end block
    ELSE    - ; separator
    WHILE   - [ start loop
    WEND    - ] end loop
    CALL    - ( function call prefix
    All other symbols are single-char tokens
    """

    def __init__(self, source: str):
        self.source = source
        self.pos = 0
        self.line = 1
        self.col = 1

    def _current(self) -> Optional[str]:
        return self.source[self.pos] if self.pos < len(self.source) else None

    def _peek(self, offset: int = 1) -> Optional[str]:
        idx = self.pos + offset
        return self.source[idx] if idx < len(self.source) else None

    def _advance(self) -> Optional[str]:
        ch = self._current()
        if ch == "\n":
            self.line += 1
            self.col = 1
        elif ch:
            self.col += 1
        self.pos += 1
        return ch

    def _skip_whitespace_and_comments(self) -> None:
        while True:
            ch = self._current()
            if ch is None:
                break
            if ch in " \t\n\r,":
                self._advance()
            elif ch == "\\" and self._peek() in ("\\", " ", "\t", None):
                # Line comment
                while self._current() is not None and self._current() != "\n":
                    self._advance()
            elif ch == "/" and self._peek() == "*":
                # Block comment
                self._advance()
                self._advance()
                while True:
                    if self._current() is None:
                        raise LexError(
                            f"Unterminated block comment at {self.line}:{self.col}"
                        )
                    if self._current() == "*" and self._peek() == "/":
                        self._advance()
                        self._advance()
                        break
                    self._advance()
            else:
                break

    def _read_number_after_first_digit(
        self, first_char: str, start_line: int, start_col: int
    ) -> Token:
        """Read a number after the first digit has been consumed."""
        num_str = first_char
        has_dot = first_char == "."

        while True:
            ch = self._current()
            if ch is None:
                break
            if ch.isdigit():
                num_str += self._advance()
            elif ch == "." and not has_dot:
                has_dot = True
                num_str += self._advance()
            else:
                break

        value = float(num_str) if has_dot else int(num_str)
        return Token("NUM", value, start_line, start_col)

    def _read_string(self) -> Token:
        """Read string literal after opening quote."""
        start_line, start_col = self.line, self.col
        self._advance()  # consume opening "
        chars = []
        escape_map = {"n": "\n", "t": "\t", "r": "\r", "\\": "\\", '"': '"'}

        while True:
            ch = self._current()
            if ch is None:
                raise LexError(f"Unterminated string at {start_line}:{start_col}")
            if ch == '"':
                self._advance()
                break
            if ch == "\\":
                self._advance()
                esc = self._current()
                if esc is None:
                    raise LexError(f"Unterminated escape at {start_line}:{start_col}")
                chars.append(escape_map.get(esc, esc))
                self._advance()
            else:
                chars.append(self._advance())

        return Token("STRING", "".join(chars), start_line, start_col)

    def _read_char(self) -> Token:
        """Read character literal after opening '."""
        start_line, start_col = self.line, self.col
        self._advance()  # consume '
        ch = self._current()
        if ch is None:
            raise LexError(f"Unterminated char at {start_line}:{start_col}")

        if ch == "\\":
            self._advance()
            esc = self._current()
            if esc is None:
                raise LexError(f"Unterminated escape at {start_line}:{start_col}")
            escape_map = {"n": "\n", "t": "\t", "r": "\r", "'": "'", "\\": "\\"}
            value = escape_map.get(esc, esc)
            self._advance()
        else:
            value = ch
            self._advance()

        return Token("CHAR", ord(value), start_line, start_col)

    def tokenize(self) -> List[Token]:
        """Tokenize source and return token list."""
        tokens = []

        # Symbol mapping for single-char tokens
        symbols = {
            "@": ("DUP", "@"),
            "#": ("SWAP", "#"),
            "$": ("DROP", "$"),
            "+": ("ADD", "+"),
            "-": ("SUB", "-"),
            "*": ("MUL", "*"),
            "/": ("DIV", "/"),
            "%": ("MOD", "%"),
            "=": ("EQ", "="),
            "<": ("LT", "<"),
            ">": ("GT", ">"),
            "~": ("NOT", "~"),
            "|": ("PRINT", "|"),
            "^": ("PRINTC", "^"),
            "?": ("INPUT", "?"),
            ":": ("STORE", ":"),
            "[": ("WHILE", "["),
            "]": ("WEND", "]"),
            "(": ("CALL", "("),
            ")": ("ENDCALL", ")"),
            "{": ("BLOCK", "{"),
            "}": ("ENDB", "}"),
            ";": ("ELSE", ";"),
            "!": ("PUSH", "!"),
            "S": ("SIN", "S"),
            "C": ("COS", "C"),
            "T": ("TAN", "T"),
            "R": ("SQRT", "R"),
            "P": ("POW", "P"),
            "F": ("FLOOR", "F"),
            "L": ("LOG", "L"),
            "E": ("EXP", "E"),
            "M": ("ABS", "M"),
            "N": ("ATAN2", "N"),
            "W": ("RAND", "W"),
            "X": ("EXIT", "X"),
            "Z": ("TIME", "Z"),
            "&": ("ALEN", "&"),
            "A": ("ALOAD", "A"),
            "a": ("ASTORE", "a"),
            "_": ("AINIT", "_"),
            "U": ("USLEEP", "U"),
        }

        while True:
            self._skip_whitespace_and_comments()
            ch = self._current()

            if ch is None:
                tokens.append(Token("EOF", None, self.line, self.col))
                break

            start_line, start_col = self.line, self.col

            # String literal
            if ch == '"':
                tokens.append(self._read_string())
                continue

            # Character literal
            if ch == "'":
                tokens.append(self._read_char())
                continue

            # Push number: ! followed by digits
            if ch == "!":
                self._advance()
                self._skip_whitespace_and_comments()
                first = self._current()
                if first is None or not (
                    first.isdigit() or first == "-" or first == "."
                ):
                    raise LexError(
                        f"Expected number after '!' at {start_line}:{start_col}"
                    )
                if first == "-":
                    self._advance()
                    next_digit = self._current()
                    if next_digit is None or not next_digit.isdigit():
                        raise LexError(
                            f"Expected digit after '-' at {start_line}:{start_col}"
                        )
                    self._advance()
                    tok = self._read_number_after_first_digit(
                        next_digit, start_line, start_col
                    )
                    tok.value = -abs(tok.value)
                elif first == ".":
                    self._advance()
                    tok = self._read_number_after_first_digit(
                        first, start_line, start_col
                    )
                else:
                    self._advance()
                    tok = self._read_number_after_first_digit(
                        first, start_line, start_col
                    )
                tokens.append(tok)
                continue

            # Digits are variable indices (push var[n])
            if ch.isdigit():
                self._advance()
                tok = self._read_number_after_first_digit(ch, start_line, start_col)
                # Integers 0-99 are variable refs
                if isinstance(tok.value, int) and tok.value <= 99:
                    tok.kind = "VAR"
                tokens.append(tok)
                continue

            # Single-char symbols
            if ch in symbols:
                self._advance()
                kind, val = symbols[ch]
                tokens.append(Token(kind, val, start_line, start_col))
                continue

            raise LexError(f"Unexpected character {ch!r} at {start_line}:{start_col}")

        return tokens

File: src/lexer.rs
//! Lexer for the sigi symbolic stack language.
//!
//! Every valid token is a single punctuation character — no alphanumeric
//! keywords.  Identifiers and store-targets use `.` and `:.` prefixes.
//! The lexer also strips C-style block comments and line comments (`\`).

use crate::ast::SourceLocation;
use std::iter::Peekable;
use std::str::Chars;
use thiserror::Error;

// ─── Lex Errors ─────────────────────────────────────────────────

#[derive(Error, Debug, PartialEq, Clone)]
pub enum LexError {
    #[error("Unterminated string at {0:?}")]
    UnterminatedString(SourceLocation),
    #[error("Unterminated escape at {0:?}")]
    UnterminatedEscape(SourceLocation),
    #[error("Unterminated char at {0:?}")]
    UnterminatedChar(SourceLocation),
    #[error("Unterminated block comment at {0:?}")]
    UnterminatedBlockComment(SourceLocation),
    #[error("Expected number after '!' at {0:?}")]
    ExpectedNumber(SourceLocation),
    #[error("Expected digit after '-' at {0:?}")]
    ExpectedDigit(SourceLocation),
    #[error("Expected name after '{0}' at {1:?}")]
    ExpectedName(String, SourceLocation),
    #[error("Unexpected character {0:?} at {1:?}")]
    UnexpectedCharacter(char, SourceLocation),
}

// ─── Token Types ─────────────────────────────────────────────────

/// Kinds of tokens the sigi language recognises.
///
/// The naming follows the symbolic character, not the semantic operation,
/// so mapping stays one-to-one with the language spec.
#[derive(Debug, Clone, PartialEq, Copy)]
pub enum TokenKind {
    NUM,
    VAR,
    IDENT,
    StoreIdent,
    CHAR,
    STRING,
    BLOCK,
    ENDB,
    ELSE,
    WHILE,
    WEND,
    CALL,
    ENDCALL,
    DUP,
    SWAP,
    DROP,
    ADD,
    SUB,
    MUL,
    DIV,
    MOD,
    EQ,
    LT,
    GT,
    NOT,
    PRINT,
    PRINTC,
    INPUT,
    STORE,
    SIN,
    COS,
    TAN,
    SQRT,
    POW,
    FLOOR,
    LOG,
    EXP,
    ABS,
    ATAN2,
    RAND,
    EXIT,
    TIME,
    ALEN,
    ALOAD,
    ASTORE,
    AINIT,
    AFREE,
    FileOpen,
    FileRead,
    FileWrite,
    FileClose,
    USLEEP,
    EOF,
}

/// A single token with its kind, optional payload, and source location.
#[derive(Debug, Clone, PartialEq)]
pub struct Token {
    pub kind: TokenKind,
    pub value: Option<TokenValue>,
    pub loc: SourceLocation,
}

/// The typed payload carried by some tokens (numbers, vars, names, chars, strings).
#[derive(Debug, Clone, PartialEq)]
pub enum TokenValue {
    Num(f64),
    Var(usize),
    Ident(String),
    Char(u8),
    String(String),
}

// ─── Lexer ───────────────────────────────────────────────────────

/// Character-level tokeniser for sigi source.
///
/// Advances through the source char-by-char, tracking line/col for
/// error reporting.  Whitespace, commas, and comments are skipped
/// silently between tokens.
#[derive(Clone)]
pub struct Lexer<'a> {
    chars: Peekable<Chars<'a>>,
    line: usize,
    col: usize,
}

impl<'a> Lexer<'a> {
    pub fn new(source: &'a str) -> Self {
        Self {
            chars: source.chars().peekable(),
            line: 1,
            col: 1,
        }
    }

    fn peek(&mut self) -> Option<char> {
        self.chars.peek().copied()
    }

    fn advance(&mut self) -> Option<char> {
        let ch = self.chars.next()?;
        if ch == '\n' {
            self.line += 1;
            self.col = 1;
        } else {
            self.col += 1;
        }
        Some(ch)
    }

    fn current_loc(&self) -> SourceLocation {
        SourceLocation {
            line: self.line,
            col: self.col,
        }
    }

    fn skip_whitespace_and_comments(&mut self) -> Result<(), LexError> {
        while let Some(ch) = self.peek() {
            if ch.is_whitespace() || ch == ',' {
                // Commas are treated as whitespace — they exist solely to
                // improve readability in dense symbolic code.
                self.advance();
            } else if ch == '\\' {
                // Line comment: skip to end of line.
                self.advance();
                while let Some(c) = self.peek() {
                    if c == '\n' {
                        break;
                    }
                    self.advance();
                }
            } else if ch == '/' && self.peek_nth(1) == Some('*') {
                // Block comment: C-style /* ... */.
                let loc = self.current_loc();
                self.advance();
                self.advance();
                while let Some(c) = self.peek() {
                    if c == '*' && self.peek_nth(1) == Some('/') {
                        self.advance();
                        self.advance();
                        break;
                    }
                    if self.advance().is_none() {
                        return Err(LexError::UnterminatedBlockComment(loc));
                    }
                }
            } else {
                break;
            }
        }
        Ok(())
    }

    fn peek_nth(&mut self, n: usize) -> Option<char> {
        self.chars.clone().nth(n)
    }

    pub fn next_token(&mut self) -> Result<Token, LexError> {
        self.skip_whitespace_and_comments()?;

        let loc = self.current_loc();
        let ch = match self.advance() {
            Some(c) => c,
            None => {
                return Ok(Token {
                    kind: TokenKind::EOF,
                    value: None,
                    loc,
                })
            }
        };

        match ch {
            '"' => self.read_string(loc),
            '\'' => self.read_char(loc),
            '!' => self.read_number(loc), // !prefix for numeric literals
            '.' => self.read_identifier(loc), // .name for named vars/functions
            ':' => {
                if self.peek() == Some('.') {
                    // :.name for store-to-named
                    self.advance();
                    self.read_store_ident(loc)
                } else {
                    Ok(Token {
                        kind: TokenKind::STORE,
                        value: None,
                        loc,
                    })
                }
            }
            c if c.is_ascii_digit() => self.read_var(c, loc), // bare digits → variable ref
            _ => self.read_symbol(ch, loc),
        }
    }

    // ─── Token Readers ───────────────────────────────────────────

    /// Read a double-quoted string, processing escape sequences.
    fn read_string(&mut self, loc: SourceLocation) -> Result<Token, LexError> {
        let mut chars = String::new();
        while let Some(ch) = self.advance() {
            if ch == '"' {
                return Ok(Token {
                    kind: TokenKind::STRING,
                    value: Some(TokenValue::String(chars)),
                    loc,
                });
            }
            if ch == '\\' {
                let esc = self.advance().ok_or(LexError::UnterminatedString(loc))?;
                chars.push(match esc {
                    'n' => '\n',
                    't' => '\t',
                    'r' => '\r',
                    '\\' => '\\',
                    '"' => '"',
                    _ => esc,
                });
            } else {
                chars.push(ch);
            }
        }
        Err(LexError::UnterminatedString(loc))
    }

    /// Read a single-quoted character literal.
    fn read_char(&mut self, loc: SourceLocation) -> Result<Token, LexError> {
        let ch = self.advance().ok_or(LexError::UnterminatedChar(loc))?;
        let value = if ch == '\\' {
            let esc = self.advance().ok_or(LexError::UnterminatedChar(loc))?;
            match esc {
                'n' => b'\n',
                't' => b'\t',
                'r' => b'\r',
                '\'' => b'\'',
                '\\' => b'\\',
                _ => esc as u8,
            }
        } else {
            ch as u8
        };
        Ok(Token {
            kind: TokenKind::CHAR,
            value: Some(TokenValue::Char(value)),
            loc,
        })
    }

    /// Read a numeric literal after the `!` prefix (e.g. `!3.14`).
    /// Negative literals use `!-n` syntax.
    fn read_number(&mut self, loc: SourceLocation) -> Result<Token, LexError> {
        let first = self.peek().ok_or(LexError::ExpectedNumber(loc))?;
        if !first.is_ascii_digit() && first != '-' && first != '.' {
            return Err(LexError::ExpectedNumber(loc));
        }

        let mut num_str = String::new();
        if first == '-' {
            num_str.push(self.advance().unwrap());
            let next = self.peek().ok_or(LexError::ExpectedDigit(loc))?;
            if !next.is_ascii_digit() {
                return Err(LexError::ExpectedDigit(loc));
            }
        }

        while let Some(c) = self.peek() {
            if c.is_ascii_digit() || c == '.' {
                num_str.push(self.advance().unwrap());
            } else {
                break;
            }
        }

        let val: f64 = num_str.parse().map_err(|_| LexError::ExpectedNumber(loc))?;
        Ok(Token {
            kind: TokenKind::NUM,
            value: Some(TokenValue::Num(val)),
            loc,
        })
    }

    /// Read a named identifier after `.` prefix.
    fn read_identifier(&mut self, loc: SourceLocation) -> Result<Token, LexError> {
        let mut name = String::new();
        while let Some(c) = self.peek() {
            if c.is_alphanumeric() || c == '_' {
                name.push(self.advance().unwrap());
            } else {
                break;
            }
        }
        if name.is_empty() {
            return Err(LexError::ExpectedName(".".to_string(), loc));
        }
        Ok(Token {
            kind: TokenKind::IDENT,
            value: Some(TokenValue::Ident(name)),
            loc,
        })
    }

    /// Read a store-target name after `:.` prefix.
    fn read_store_ident(&mut self, loc: SourceLocation) -> Result<Token, LexError> {
        let mut name = String::new();
        while let Some(c) = self.peek() {
            if c.is_alphanumeric() || c == '_' {
                name.push(self.advance().unwrap());
            } else {
                break;
            }
        }
        if name.is_empty() {
            return Err(LexError::ExpectedName(":.".to_string(), loc));
        }
        Ok(Token {
            kind: TokenKind::StoreIdent,
            value: Some(TokenValue::Ident(name)),
            loc,
        })
    }

    /// Read a variable reference (bare digits 0--99).  Larger numbers
    /// are returned as NUM to allow non-var constant push.
    fn read_var(&mut self, first: char, loc: SourceLocation) -> Result<Token, LexError> {
        let mut num_str = first.to_string();
        while let Some(c) = self.peek() {
            if c.is_ascii_digit() {
                num_str.push(self.advance().unwrap());
            } else {
                break;
            }
        }
        let val: usize = num_str.parse().unwrap();
        Ok(Token {
            kind: if val <= 99 {
                TokenKind::VAR
            } else {
                TokenKind::NUM
            },
            value: Some(TokenValue::Var(val)),
            loc,
        })
    }

    /// Map a single punctuation character to its `TokenKind`.
    /// This is the core symbol table of the sigi language.
    fn read_symbol(&mut self, ch: char, loc: SourceLocation) -> Result<Token, LexError> {
        let kind = match ch {
            '@' => TokenKind::DUP,
            '#' => TokenKind::SWAP,
            '$' => TokenKind::DROP,
            '+' => TokenKind::ADD,
            '-' => TokenKind::SUB,
            '*' => TokenKind::MUL,
            '/' => TokenKind::DIV,
            '%' => TokenKind::MOD,
            '=' => TokenKind::EQ,
            '<' => TokenKind::LT,
            '>' => TokenKind::GT,
            '~' => TokenKind::NOT,
            '|' => TokenKind::PRINT,
            '^' => TokenKind::PRINTC,
            '?' => TokenKind::INPUT,
            '[' => TokenKind::WHILE,
            ']' => TokenKind::WEND,
            '(' => TokenKind::CALL,
            ')' => TokenKind::ENDCALL,
            '{' => TokenKind::BLOCK,
            '}' => TokenKind::ENDB,
            ';' => TokenKind::ELSE,
            'S' => TokenKind::SIN,
            'C' => TokenKind::COS,
            'T' => TokenKind::TAN,
            'R' => TokenKind::SQRT,
            'P' => TokenKind::POW,
            'F' => TokenKind::FLOOR,
            'L' => TokenKind::LOG,
            'E' => TokenKind::EXP,
            'M' => TokenKind::ABS,
            'N' => TokenKind::ATAN2,
            'W' => TokenKind::RAND,
            'X' => TokenKind::EXIT,
            'Z' => TokenKind::TIME,
            '&' => TokenKind::ALEN,
            'A' => TokenKind::ALOAD,
            'a' => TokenKind::ASTORE,
            '_' => TokenKind::AINIT,
            'K' => TokenKind::AFREE,
            'O' => TokenKind::FileOpen,
            'G' => TokenKind::FileRead,
            'H' => TokenKind::FileWrite,
            'Y' => TokenKind::FileClose,
            'U' => TokenKind::USLEEP,
            _ => return Err(LexError::UnexpectedCharacter(ch, loc)),
        };
        Ok(Token {
            kind,
            value: None,
            loc,
        })
    }
}

// ─── Iterator Adapter ────────────────────────────────────────────

/// Wrap `next_token` as a standard iterator, suppressing the EOF sentinel.
impl<'a> Iterator for Lexer<'a> {
    type Item = Result<Token, LexError>;

    fn next(&mut self) -> Option<Self::Item> {
        match self.next_token() {
            Ok(token) => {
                if token.kind == TokenKind::EOF {
                    None
                } else {
                    Some(Ok(token))
                }
            }
            Err(e) => Some(Err(e)),
        }
    }
}

// ─── Tests ───────────────────────────────────────────────────────

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_lex_simple() {
        let mut lexer = Lexer::new("!5 @ +");
        let t1 = lexer.next_token().unwrap();
        assert_eq!(t1.kind, TokenKind::NUM);
        assert_eq!(t1.value, Some(TokenValue::Num(5.0)));
        let t2 = lexer.next_token().unwrap();
        assert_eq!(t2.kind, TokenKind::DUP);
        let t3 = lexer.next_token().unwrap();
        assert_eq!(t3.kind, TokenKind::ADD);
        let t4 = lexer.next_token().unwrap();
        assert_eq!(t4.kind, TokenKind::EOF);
    }
}
